pkg/util/fxutil/provide_comp_test.go

```diff
@@ -0,0 +1,414 @@
+		})
+		return providesService{First: &firstImpl{}}
+	}
+	start := func(one FirstComp) {
+		counter += 2
+	}
+	module := Component(ProvideComponentConstructor(NewAgentComponent))
+	Test[FirstComp](t, fx.Invoke(start), module)
+	// ensures that both the OnStart hook and entry point are called
+	require.Equal(t, 3, counter)
+}
+
+func assertNoCtorError(t *testing.T, arg fx.Option) {
+	app := fx.New(arg)
+	err := app.Err()
+	if err != nil {
+		t.Fatalf("expected no error, instead got %s", err)
+	}
+}
+
+func assertIsSingleError(t *testing.T, arg fx.Option, errMsg string) {
+	app := fx.New(arg)
+	err := app.Err()
+	if err == nil {
+		t.Fatalf("expected an error, instead got %v of type %T", arg, arg)
+	} else if err.Error() != errMsg {
+		t.Fatalf("errror mismatch, expected %v, got %v", errMsg, err.Error())
+	}
+}
+
+func getFieldNames(it interface{}) []string {
+	var typ reflect.Type
+	if t, ok := it.(reflect.Type); ok {
+		typ = t
+	} else {
+		typ = reflect.TypeOf(it)
+	}
+	if typ.Kind() != reflect.Struct {
+		return nil
+	}
+	fields := make([]string, 0, typ.NumField())
+	for i := 0; i < typ.NumField(); i++ {
+		fields = append(fields, typ.Field(i).Name)
+	}
+	return fields
+}
+
+// sample test data follows
+
+// FirstComp is an interface for a Component, it doesn't do anything
+
+type FirstComp interface {
+	String() string
+}
+
+type firstImpl struct{}
+
+func (m *firstImpl) String() string {
+	return "1st"
+}
+
+// SecondComp is an interface for a Component, it depends on FirstComp for construction
+
+type SecondComp interface {
+	Second() string
+}
+
+type secondImpl struct {
+	First FirstComp
+}
+
+func (m *secondImpl) Second() string {
+	return "2nd"
+}
+
+// provides1 provides 1 component using a composite struct
+
+type provides1 struct {
+	First FirstComp
+}
+
+type fxProvides1 struct {
+	fx.Out
+	First FirstComp
+}
+
+// provides2 provides a different component
+
+type provides2 struct {
+	Second SecondComp
+}
+
+// provides3 provides a composite struct that embeds another
+
+type provides3 struct {
+	A Apple
+	B Banana
+	C CherryProvider
+}
+
+type Apple struct {
+	D Donut
+}
+
+type Banana interface{}
+
+type bananaImpl struct {
+	Color string
+}
+
+type CherryProvider struct {
+	compDef.Out
+	E Egg
+}
+
+type Donut struct {
+	X int
+}
+
+type Egg struct {
+	Y int
+}
+
+// provides4 contains two different embedding structs
+
+type provides4 struct {
+	C CherryProvider
+	F FruitProvider
+}
+
+type FruitProvider struct {
+	compDef.Out
+	Z int
+}
+
+// requires1 requires 1 component using a composite struct
+
+type requires1 struct {
+	First FirstComp
+}
+
+// requires2 requires a different component
+
+type requires2 struct {
+	Second SecondComp
+}
+
+// requiresLc uses Lifecycles
+
+type requiresLc struct {
+	Lc compDef.Lifecycle
+}
+
+type providesService struct {
+	First FirstComp
+}
+
+// fxAwareReqs is an fx-aware requires struct
+
+type fxAwareReqs struct {
+	fx.In
+	A Apple
+}
```

Comment:
Would make sense to place this before the test functions. It is easier to navigate the file 😄

Reply:
Does having the data before the tests make it easier to navigate? It seems the same either way, but I could be convinced.